bigquery_etl/routine/parse_routine.py

```diff
@@ -272,25 +289,33 @@ def routine_usage_definitions(text, project, raw_routines=None):
     ]
 
 
-def sub_local_routines(test, project, raw_routines=None):
+def sub_local_routines(test, project, raw_routines=None, stored_procedure_test=False):
     """
-    Transform persistent UDFs into temporary UDFs.
+    Transform persistent UDFs into temporary UDFs for UDF tests.
 
-    Use generic dataset for stored procedures.
+    Use generic dataset for stored procedure tests.
     """
+    if raw_routines is None:
+        raw_routines = read_routine_dir()
+
     sql = prepend_routine_usage_definitions(test, project, raw_routines)
-    sql = sub_persistent_udf_names_as_temp(sql)
-    routines = get_routines(project)
 
-    for routine in routines:
-        if routine["name"] in sql:
-            replace_name = routine["name"].replace(".", "_")
-            if not routine["is_udf"]:
-                replace_name = GENERIC_DATASET + "." + replace_name
-            sql = sql.replace(f'{routine["project"]}.{routine["name"]}', replace_name)
-            sql = sql.replace(f'{routine["name"]}', replace_name)
+    for name, routine in raw_routines.items():
+        if name in sql:
+            for defn in routine.definitions:
+                match = PERSISTENT_UDF_RE.match(defn)
+                dataset, name = match.group(1), match.group(2)
+                replace_name = f"{dataset}_{name}"
+                if stored_procedure_test:
+                    replace_name = f"{GENERIC_DATASET}.{replace_name}"
+                sql = re.sub(
+                    fr"`?(?:`?{routine.project}`?\.)?{dataset}.{name}`?",
```

Comment:
What is this doing, with all of the `?`? Conditionally setting the project if it exists?

Reply:
right - may or may not have backticks, may or may not have backticks around the project name, the project name may or may not exist.